pkg/controller/controllerutil/controllerutil.go

```diff
@@ -168,3 +169,61 @@ func mutate(f MutateFn, key client.ObjectKey, obj runtime.Object) error {
 
 // MutateFn is a function which mutates the existing object into it's desired state.
 type MutateFn func() error
+
+// AddFinalizer accepts a metav1 object and adds the provided finalizer if not present.
+func AddFinalizer(o metav1.Object, finalizer string) {
+	newFinalizers := AddString(o.GetFinalizers(), finalizer)
+	o.SetFinalizers(append(newFinalizers, finalizer))
+}
+
+// AddFinalizerIfPossible tries to convert a runtime object to a metav1 object and add the provided finalizer.
+// It returns an error if the provided object cannot provide an accessor.
+func AddFinalizerIfPossible(o runtime.Object, finalizer string) error {
+	m, err := meta.Accessor(o)
+	if err != nil {
+		return err
+	}
+	AddFinalizer(m, finalizer)
+	return nil
+}
+
+// RemoveFinalizer accepts a metav1 object and removes the provided finalizer if present.
+func RemoveFinalizer(o metav1.Object, finalizer string) {
+	newFinalizers := RemoveString(o.GetFinalizers(), finalizer)
+	o.SetFinalizers(newFinalizers)
+}
+
+// RemoveFinalizerIfPossible tries to convert a runtime object to a metav1 object and remove the provided finalizer.
+// It returns an error if the provided object cannot provide an accessor.
+func RemoveFinalizerIfPossible(o runtime.Object, finalizer string) error {
+	m, err := meta.Accessor(o)
+	if err != nil {
+		return err
+	}
+	RemoveFinalizer(m, finalizer)
+	return nil
+}
+
+// AddString returns a []string with s appended if it is not already found in the provided slice.
+func AddString(slice []string, s string) []string {
+	newSlice := make([]string, 0)
```

Comment:
No reason to make a copy here:

```go
for i, item := range slice {
  if item == s {
     return slice
  }
}
return append(slice, s)
```

Reply:
also make these private

Reply:
i was wondering if there was a reason the k/k version made a copy, I stared at it a bit and decided no => already reduce to what you posted